### PR TITLE
fix(2425): Remove join builds in created state

### PR DESCRIPTION
**Pull request by mchengat**
## Context

For a join build in CREATED status, only a successful parent build can delete it. This leads to a situation that if the last build in join relation ship fails, then it's not cleaned up.

## Objective

Will remove child jobs in CREATED state of failed parent jobs

## References

https://github.com/screwdriver-cd/screwdriver/issues/2425

## License
I confirm that this contribution is made under a BSD license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: plugins/builds/update.js [hunk 1]
```diff
@@ -254,6 +254,20 @@ module.exports = () => ({
                                 // Don't further trigger pipeline if intented to skip further jobs
 
                                 if (newBuild.status !== 'SUCCESS' || skipFurther) {
+                                    // Check for failed jobs and remove any child jobs in created state
+                                    if (newBuild.status === 'FAILURE') {
+                                        return removeJoinChild(
```
Comment (jithine) [inline plugins/builds/update.js:259]: replace with

```
try {
await removeJoinChild(...);
} ...
```
no need for the then handler and return.

File: plugins/builds/index.js [hunk 2]
```diff
@@ -657,6 +697,75 @@ async function createJoinObject(nextJobs, current, eventFactory) {
 const buildsPlugin = {
     name: 'builds',
     async register(server, options) {
+        /**
+         * Remove the child jobs of the current job in CREATED state
```
Comment (jithine) [inline plugins/builds/index.js:701]: `child jobs` -> `downstream jobs`

File: plugins/builds/index.js [hunk 3]
```diff
@@ -657,6 +697,75 @@ async function createJoinObject(nextJobs, current, eventFactory) {
...
+                    jobName: nextJobName,
+                    event: current.event,
+                    parentBuilds
+                };
+
+                return getInternalBuild(internalBuildConfig);
+            };
+
+            for (const pid of Object.keys(pipelineJoinData)) {
+                if (+pid === current.pipeline.id) {
```
Comment (jithine) [inline plugins/builds/index.js:748]: how about external pipelines to handle remote join?

File: plugins/builds/index.js [hunk 4]
```diff
@@ -657,6 +697,75 @@ async function createJoinObject(nextJobs, current, eventFactory) {
...
+                    joinObj,
+                    current,
+                    nextJobName
+                });
+                const internalBuildConfig = {
+                    jobFactory,
+                    buildFactory,
+                    pipelineId: current.pipeline.id,
+                    jobName: nextJobName,
+                    event: current.event,
```
Comment (jithine) [inline plugins/builds/index.js:740]: this could be modified to fetch both internal & external pipelines

File: plugins/builds/update.js [hunk 5]
```diff
@@ -254,6 +254,22 @@ module.exports = () => ({
...
+                                            await removeJoinChild(
+                                                {
+                                                    pipeline,
+                                                    job,
+                                                    build: newBuild
+                                                },
+                                                request.server.app
+                                            );
+                                        } catch (err) {
+                                            throw boom.internal(`failed to remove created child jobs, ${err}`);
```
Comment (tkyi) [inline plugins/builds/update.js:269]: should this throw err or just fail silently?

File: plugins/builds/index.js [hunk 6]
```diff
@@ -657,6 +657,87 @@ async function createJoinObject(nextJobs, current, eventFactory) {
...
+                }
+
+                return null;
+            };
+
+            const delBuild = buildToDel => {
+                if (buildToDel && buildToDel.status === 'CREATED') {
+                    buildToDel.remove();
+                }
+            };
```
Comment (tkyi) [inline plugins/builds/index.js:709]: prefer these methods moved outside for better jsdocs and things

File: plugins/builds/index.js [hunk 7]
```diff
@@ -657,6 +657,87 @@ async function createJoinObject(nextJobs, current, eventFactory) {
...
+                chainPR: pipeline.chainPR
+            });
+
+            const pipelineJoinData = await createJoinObject(nextJobsTrigger, current, eventFactory);
+
+            // triggerFetchBuild fetches the build details for both internal and external
+            const triggerFetchBuild = async (nextJobName, joinObj, eventId) => {
+                const nextJob = joinObj.jobs[nextJobName];
+
+                const internalBuildConfig = {
```
Comment (tkyi) [inline plugins/builds/index.js:693]: maybe rename var since not just for internal

**Comment by tkyi** [comment]
No new tests?

File: plugins/builds/index.js [hunk 8]
```diff
@@ -65,6 +65,37 @@ function getExternalEvent(currentBuild, pipelineId, eventFactory) {
...
+
+/**
+ * Delete a build
+ * @method delBuild
+ * @param  {Object}   buildToDel                  build object to delete
+ * @return {Promise}
+ * */
+async function delBuild(buildToDel) {
+    if (buildToDel && buildToDel.status === 'CREATED') {
+        buildToDel.remove();
```
Comment (tkyi) [inline plugins/builds/index.js:95]:
```suggestion
        return buildToDel.remove();
```

File: plugins/builds/index.js [hunk 9]
```diff
@@ -65,6 +65,37 @@ function getExternalEvent(currentBuild, pipelineId, eventFactory) {
...
+ * Fetch builds with given params
+ * @method triggerFetchBuild
+ * @param  {Object}   config                    Parameters of build to fetch
+ * @param  {Factory}  config.buildFactory       Build Factory
+ * @param  {Number}   config.jobID              Job ID
+ * @param  {Number}   config.eventID            Event ID
+ * @return {Promise}
+ * */
+async function triggerFetchBuild(config) {
+    const { jobID, eventID, buildFactory } = config;
```
Comment (tkyi) [inline plugins/builds/index.js:78]:
```suggestion
async function triggerFetchBuild({ jobID, eventID, buildFactory }) {
```

File: plugins/builds/index.js [hunk 10]
```diff
@@ -65,6 +65,37 @@ function getExternalEvent(currentBuild, pipelineId, eventFactory) {
...
+    return buildFactory.get(buildConfig);
+}
+
+/**
+ * Delete a build
+ * @method delBuild
+ * @param  {Object}   buildToDel                  build object to delete
+ * @return {Promise}
+ * */
+async function delBuild(buildToDel) {
```
Comment (jithine) [inline plugins/builds/index.js:93]: call this `deleteBuild`

File: plugins/builds/index.js [hunk 11]
```diff
@@ -65,6 +65,37 @@ function getExternalEvent(currentBuild, pipelineId, eventFactory) {
...
+/**
+ * Fetch builds with given params
+ * @method triggerFetchBuild
+ * @param  {Object}   config                    Parameters of build to fetch
+ * @param  {Factory}  config.buildFactory       Build Factory
+ * @param  {Number}   config.jobID              Job ID
+ * @param  {Number}   config.eventID            Event ID
+ * @return {Promise}
+ * */
+async function triggerFetchBuild(config) {
```
Comment (jithine) [inline plugins/builds/index.js:77]: why is this function needed ? it's just wrapping a single line, make this inline where it's used

File: plugins/builds/update.js [hunk 12]
```diff
@@ -254,6 +254,18 @@ module.exports = () => ({
                                 // Don't further trigger pipeline if intented to skip further jobs
 
                                 if (newBuild.status !== 'SUCCESS' || skipFurther) {
+                                    // Check for failed jobs and remove any child jobs in created state
+                                    if (newBuild.status === 'FAILURE') {
+                                        await removeJoinChild(
```
Comment (jithine) [inline plugins/builds/update.js:259]: call this removeJoinBuilds

File: plugins/builds/index.js [hunk 13]
```diff
@@ -662,6 +693,70 @@ async function createJoinObject(nextJobs, current, eventFactory) {
 const buildsPlugin = {
     name: 'builds',
     async register(server, options) {
+        /**
+         * Remove the downstream jobs of the current job in CREATED state
```
Comment (jithine) [inline plugins/builds/index.js:697]: `Remove builds for downstream jobs of current job ...`

File: plugins/builds/index.js [hunk 14]
```diff
@@ -662,6 +693,70 @@ async function createJoinObject(nextJobs, current, eventFactory) {
...
+         * @method removeJoinChild
+         * @param {Object}      config              Configuration object
+         * @param {Pipeline}    config.pipeline     Current pipeline
+         * @param {Job}         config.job          Current job
+         * @param {Build}       config.build        Current build
+         * @param {String}      config.username     Username
+         * @param {String}  app                      Server app object
+         * @return {Promise}                        Resolves to the removed build or null
+         */
+        server.expose('removeJoinChild', async (config, app) => {
```
Comment (jithine) [inline plugins/builds/index.js:707]: call this `removeJoinBuilds` it's more semantic

File: plugins/builds/index.js [hunk 15]
```diff
@@ -662,6 +693,70 @@ async function createJoinObject(nextJobs, current, eventFactory) {
...
+                    try {
+                        const nextJob = pipelineJoinData[pid].jobs[nextJobName];
+
+                        buildConfig.jobID = nextJob.id;
+                        if (!isExternal) {
+                            buildConfig.eventID = event.id;
+                        } else {
+                            buildConfig.eventID = pipelineJoinData[pid].event.id;
+                        }
+                        buildToDel = await triggerFetchBuild(buildConfig);
```
Comment (jithine) [inline plugins/builds/index.js:743]: call this `buildToDelete`

File: plugins/builds/index.js [hunk 16]
```diff
@@ -662,6 +693,70 @@ async function createJoinObject(nextJobs, current, eventFactory) {
...
+
+                        if (buildToDel !== null) {
```
Comment (jithine) [inline plugins/builds/index.js:745]: this if check is redundant you are already checking for this in `delBuild`

**Review by jithine — APPROVED** [review]

**Pull request merged**